### DISCUSSION TITLE
with expression + interfaces

### DISCUSSION THREAD

**Discussion by JanEggers** [dotnet/csharplang#4158] [Q&A] [answered]
right now im having a bunch of interfaces describing my model. I would like to convert the implementations into records.

so the interface gets init properties instead of readonly properties. 

but how do I express that the object is withable in the interface?

**Comment by Joe4evr** [comment, marked as answer]
You can't right now.

For the next version of C#, the language team intends to add `[Factory]`: an attribute that indicates a method return is always a fresh instance, so it's okay to put an object initializer after such a call. If you then define a method named `With` that takes no parameters and returns the same object type, the compiler will use that pattern to allow `with` as a keyword on instances of that object.
```cs
public interface IMyModel
{
    [Factory]
    IMyModel With();
}
```
[See here.](https://youtu.be/fIXujTOA6BE?t=3023)

**Reply by 333fred** [reply]
How we intend to add either factories or the general "withable" constraint is still _very_ much undecided. I wouldn't say anything about what we intend beyond a simple we intend to support this.
